bioio/tests/conftest.py

```diff
@@ -61,51 +50,277 @@ def da_random_from_shape(
+        return list(self._exts)
+
+    def get_reader(self) -> Type[BaseReader]:
+        return self._reader_cls
+
+
+def _make_reader_class(spec: TestPluginSpec) -> Type[BaseReader]:
+    """
+    Create Reader subclass for synthetic plugin.
+    """
+
+    class TestReader(BaseReader):
+        NAME = spec.name
+
+        def __init__(
+            self,
+            image: Any,
+            fs_kwargs: Optional[dict[str, Any]] = None,
+            **kwargs: Any,
+        ) -> None:
+            fs_kwargs = fs_kwargs or {}
+
+            # If requested, enforce anon=True (for S3 tests)
+            if spec.check_anon_in_init and not fs_kwargs.get("anon", False):
+                raise bioio_base.exceptions.UnsupportedFileFormatError(
+                    "test", "test", spec.fail_message
+                )
+
+        @property
+        def scenes(self) -> list[str]:
+            return ["Scene 0"]
+
+        def _read_immediate(self, *args: Any, **kwargs: Any) -> Any:
+            raise NotImplementedError("_read_immediate not used in tests")
+
+        def _read_delayed(self, *args: Any, **kwargs: Any) -> Any:
+            raise NotImplementedError("_read_delayed not used in tests")
+
+        @classmethod
+        def _is_supported_image(
+            cls,
+            image: Any,
+            fs_kwargs: Optional[dict[str, Any]] = None,
+            **kwargs: Any,
+        ) -> bool:
+            if spec.fail_on_is_supported:
+                raise TypeError(spec.fail_message)
+            return True
+
+    TestReader.__name__ = f"{spec.name.title().replace('-', '_')}Reader"
+    return TestReader
+
+
+def _install_ephemeral_module(spec: TestPluginSpec) -> str:
+    """
+    Create a synthetic module in sys.modules exposing ReaderMetadata.
+    """
+    module_name = spec.module_name or f"bioio_test_plugins.{spec.name}"
+    mod = types.ModuleType(module_name)
+
+    reader_cls = _make_reader_class(spec)
+    reader_metadata = _TestReaderMetadata(
+        name=spec.name,
+        exts=spec.supported_extensions,
+        reader_cls=reader_cls,
+    )
+
+    setattr(mod, "Reader", reader_cls)
+    setattr(mod, "ReaderMetadata", reader_metadata)
+
+    sys.modules[module_name] = mod
+    return module_name
+
+
+def _make_entry_points(specs: Iterable[TestPluginSpec]) -> List[EntryPoint]:
+    eps: List[EntryPoint] = []
+    for spec in specs:
+        module_name = _install_ephemeral_module(spec)
+        ep = EntryPoint(
+            name=spec.name,
+            group="bioio.readers",
+            value=module_name,
         )
+        eps.append(ep)
+    return eps
 
 
 @pytest.fixture
-def dummy_plugin() -> typing.Generator[str, None, None]:
-    with InstallPackage(package_path=DUMMY_PLUGIN_PATH, package_name=DUMMY_PLUGIN_NAME):
-        yield DUMMY_PLUGIN_NAME
+def plugin_factory(
+    monkeypatch: MonkeyPatch,
+) -> Callable[[Iterable[TestPluginSpec]], List[EntryPoint]]:
+    """
+    Factory fixture to install synthetic plugins.
```

Comment:
this needs way more commenting right here about what it does.  it just needs to say that it is creating test plugins as fake modules with entrypoints so that after calling the factory method, the plugins associated with the input TestPluginSpecs are actually "installed" and usable as bioio readers in the tests.